prusa/link/cameras/encoders.py

```diff
@@ -0,0 +1,399 @@
+class BufferDetails:
+    """A structure to encapsulate buffer info needed for encoding"""
+
+    def __init__(self, file_descriptor, length, offset):
+        self.file_descriptor = file_descriptor
+        self.length = length
+        self.offset = offset
+        self.mmap = mmap.mmap(fileno=self.file_descriptor,
+                              length=self.length,
+                              offset=self.offset)
+
+    def __del__(self):
+        try:
+            self.mmap.close()
+        except AttributeError:
+            pass
+
+
+def get_appropriate_encoder(resolution, pixel_format, use_mmap=False):
+    """Returns the appropriate encoder based on stream parameters"""
+    max_resolution = max(resolution.width, resolution.height)
+    if pixel_format == v4l2.V4L2_PIX_FMT_MJPEG:
+        return PassthroughEncoder()
+    if not MJPEGEncoder.is_available():
+        return JPEGEncoder()
+    if max_resolution > MJPEGEncoder.WIDTH_LIMIT:
+        return JPEGEncoder()
+    encoder = MJPEGEncoder()
+    if use_mmap:
+        # Switch to a type that copies data instead of trying to use
+        # a foreign buffer
+        encoder.ingest_buffer_memory = v4l2.V4L2_MEMORY_MMAP
+    return encoder
+
+
+class Encoder:
+    """A base class for encoders"""
+
+    def __init__(self):
+        """Set all parameters encoder needs after calling init"""
+        self.width = 0
+        self.height = 0
+        self.stride = 0
+        self.fps = 30
+        self._quality = Quality.HIGH
+
+        # Information about the buffer from which to read
+        self.source_details = None
+
+    def start(self):
+        """Initializes the encoder"""
+
+    def stop(self):
+        """Stops the encoder"""
+
+    @property
+    def quality(self):
+        """Gets the quality"""
+        return self._quality
+
+    @quality.setter
+    def quality(self, quality=Quality.HIGH):
+        """An entry point for other parameters dependant on quality"""
+        self._quality = quality
+
+    @abc.abstractmethod
+    def encode(self, bytes_used: int) -> bytes:
+        """Encode here, return bytes"""
+
+
+class MJPEGEncoder(Encoder):
+    """Encoder using the MJPEG Encoder on the Raspberry Pi through V4L2
+
+    Glossary:
+    SOURCE means foreign object like a buffer we copy data from
+    INGEST means our own data structure with raw data (V4L2 name: Output)
+    CODED means the structure with compressed data (V4L2 name: Capture)
+    """
+    WIDTH_LIMIT = 1920
+
+    # These are suggested bitrates for 1080p30 in Mbps
+    BITRATE_TABLE = {Quality.VERY_LOW: 6,
+                     Quality.LOW: 12,
+                     Quality.MEDIUM: 18,
+                     Quality.HIGH: 27,
+                     Quality.VERY_HIGH: 45}
+
+    # Use only one buffer, so no indexes need to exist
+    BUFFER_INDEX = 0
+
+    @staticmethod
+    @functools.cache
+    def is_available():
+        """Figures whether we can do hardware decode or not"""
+        if not os.path.exists(PATH):
```

Comment:
BTW: may be better place for HW MJPEGEncoder path is class, why you didn't use class attribute?